File: week 4/sorting/sorting.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::vector;
using std::swap;

vector<int> partition3(vector<int> &a, int l, int r) {
	int x = a[l];
    int m1 = l;
	int m2 = r;
	int i = l;
    while(i <= m2){
		if (a[i] < x) {
			swap(a[i], a[m1]);
			m1++;
			i++;
		}
		else if (a[i] > x) {
			swap(a[i], a[m2]);
			m2--;
		}
		else
			i++;
    }
	vector<int> m = {m1, m2};
    return m;
}

int partition2(vector<int> &a, int l, int r) {
  int x = a[l];
  int j = l;
  for (int i = l + 1; i <= r; i++) {
    if (a[i] <= x) {
      j++;
      swap(a[i], a[j]);
    }
  }
  swap(a[l], a[j]);
  return j;
}

void randomized_quick_sort(vector<int> &a, int l, int r) {
  if (l >= r) {
    return;
  }

  int k = l + rand() % (r - l + 1);
  swap(a[l], a[k]);
  // int m = partition2(a, l, r);

  // randomized_quick_sort(a, l, m - 1);
  // randomized_quick_sort(a, m + 1, r);
  vector<int> m = partition3(a, l, r);

  randomized_quick_sort(a, l, m[0] - 1);
  randomized_quick_sort(a, m[1] + 1, r);
}

int main() {
  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); ++i) {
    std::cin >> a[i];
  }
  randomized_quick_sort(a, 0, a.size() - 1);
  for (size_t i = 0; i < a.size(); ++i) {
    std::cout << a[i] << ' ';
  }
}

File: week 4/sorting/sorting.py
import random

def partition3(a, l, r):
    x = a[l]
    m1 = l
    m2 = r
    i = l
    while(i <= m2):
        if (a[i] < x):
            a[i], a[m1] = a[m1], a[i]
            m1 += 1
            i += 1
        elif (a[i] > x):
			a[i], a[m2] = a[m2], a[i]
			m2 -= 1
        else:
			i += 1
	m = [m1, m2]
    return m

def partition2(a, l, r):
    x = a[l]
    j = l;
    for i in range(l + 1, r + 1):
        if a[i] <= x:
            j += 1
            a[i], a[j] = a[j], a[i]
    a[l], a[j] = a[j], a[l]
    return j


def randomized_quick_sort(a, l, r):
    if l >= r:
        return
    k = random.randint(l, r)
    a[l], a[k] = a[k], a[l]
    #use partition3
    # m = partition2(a, l, r)
    # randomized_quick_sort(a, l, m - 1)
    # randomized_quick_sort(a, m + 1, r)
    m = partition3(a, l, r)
    randomized_quick_sort(a, l, m[0] - 1)
    randomized_quick_sort(a, m[1] + 1, r)


if __name__ == '__main__':
    n = int(input())
    a = [0] * n
    a = list(map(int, raw_input().split()))
    randomized_quick_sort(a, 0, n - 1)
    for x in a:
        print str(x) + ' ',